main.go

```diff
@@ -0,0 +1,706 @@
+	lgr.exitIf(dsName == "", "Missing device service name.")
+	metadataURI.Path = "/api/v1/device/servicename/" + dsName
+	deviceNames, err := llrp.GetDevices(metadataURI.String(), http.DefaultClient)
+	lgr.exitIfErr(err, "Failed to get existing device names.", lg{"path", metadataURI.String()})
+	for _, name := range deviceNames {
+		lgr.exitIfErr(defaultGrp.AddReader(devService, name),
+			"Failed to setup device.", lg{"device", name})
+	}
+
+	app := inventoryApp{
+		lgr:          lgr,
+		edgexSdk:     edgexSdk,
+		defaultGrp:   defaultGrp,
+		devService:   devService,
+		snapshotReqs: make(chan snapshotDest),
+		reports:      make(chan reportData),
+	}
+
+	// routes
+	for _, rte := range []struct {
+		path, method string
+		f            http.HandlerFunc // of course the EdgeX SDK doesn't take a http.Handler...
+	}{
+		{"/", http.MethodGet, func(w http.ResponseWriter, r *http.Request) {
+			http.ServeFile(w, r, "static/html/index.html")
+		}},
+		{"/api/v1/readers", http.MethodGet, func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			if err := app.defaultGrp.ListReaders(w); err != nil {
+				lgr.Error("Failed to write readers list.", "error", err.Error())
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		}},
+		{"/api/v1/inventory/snapshot", http.MethodGet,
+			func(w http.ResponseWriter, req *http.Request) {
+				w.Header().Set("Content-Type", "application/json")
+				if err := app.writeInventorySnapshot(w); err != nil {
+					app.lgr.Error("Failed to write inventory snapshot.", "error", err.Error())
+					w.WriteHeader(http.StatusInternalServerError)
+				}
+			},
+		},
+		{"/api/v1/command/reading/start", http.MethodPost,
+			func(w http.ResponseWriter, req *http.Request) {
+				if err := app.defaultGrp.StartAll(devService); err != nil {
+					lgr.Error("Failed to StartAll.", "error", err.Error())
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+			},
+		},
+		{"/api/v1/command/reading/stop", http.MethodPost,
+			func(w http.ResponseWriter, req *http.Request) {
+				if err := app.defaultGrp.StopAll(devService); err != nil {
+					lgr.Error("Failed to StopAll.", "error", err)
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+			},
+		},
+		{"/api/v1/behaviors/{name}", http.MethodGet,
+			func(w http.ResponseWriter, req *http.Request) {
+				rv := mux.Vars(req)
+				bName := rv["name"]
+				// Currently, only "default" is supported.
+				if bName != "default" {
+					lgr.Error("Request to GET unknown behavior.", "name", bName)
+					if _, err := w.Write([]byte("Invalid behavior name.")); err != nil {
+						lgr.Error("Error writing failure response.", "error", err)
+					}
+					w.WriteHeader(http.StatusNotFound)
+					return
+				}
+
+				data, err := json.Marshal(app.defaultGrp.Behavior())
+				if err != nil {
+					lgr.Error("Failed to marshal behavior.", "error", err)
+					w.WriteHeader(http.StatusInternalServerError)
+					return
```

Comment:
Should write the error message as response? Here and other REST handlers.

Reply:
Typically as part of secure coding practices we prefer to return as little error information as possible to the client to avoid potentially leaking information that could be used by attackers. In this case I think it may be considered safe to return just the `"Failed to marshal behavior."` part without the actual error.

Does that sound ok?

Reply:
The question is do those `secure coding practices` apply to open source SW where potential attacker has fully access to the public code? I think not.

I'll bring this up in the Core WG meeting this morning so we can get consensus for the community. @jpwhitemn please add to agenda.

Reply:
We discuss this in today's Core WG meeting and it was agreed that these `secure coding practices we prefer to return as little error information as possible to the client` don't apply.

Please log detailed error message and return the detailed error message in the HTTP responses. THX!